lib/mjsonwp/routes.js

```diff
@@ -22,7 +24,23 @@ const METHOD_MAP = {
   },
   '/wd/hub/session/:sessionId/timeouts': {
     GET: {command: 'getTimeouts'}, // W3C route
-    POST: {command: 'timeouts', payloadParams: {required: ['type', 'ms']}}
+    POST: {command: 'timeoutsW3C', payloadParams: {
+      validate: (jsonObj, protocol) => {
+        if (protocol === BaseDriver.DRIVER_PROTOCOL.W3C) {
+          if (!util.hasValue(jsonObj.script) && !util.hasValue(jsonObj.pageLoad) && !util.hasValue(jsonObj.implicit)) {
+            return 'W3C protocol expects one or more of script, pageLoad or implicit';
+          }
+        } else {
+          if (!util.hasValue(jsonObj.type) || !util.hasValue(jsonObj.ms)) {
+            return 'MJSONWP protocol requires type and ms';
+          }
+        }
+      },
+      optional: ['type', 'ms', 'script', 'pageLoad', 'implicit']},
+      makeArgs: (jsonObj) => {
+        return [jsonObj.type, jsonObj.ms, jsonObj.script, jsonObj.pageLoad, jsonObj.implicit];
```

Comment:
here you can, actually, decide which params should be passed to the command handler based on their presence, which means we might just send an object containing 

```
{type, ms}
```

for JSONWP or 

```
script, pageLoad, implicit
```

for W3C.

I don't like methods with 5 arguments. @imurchie What do you think?

Reply:
What about https://github.com/appium/appium-base-driver/pull/164/commits/4e571cbf153078a05f6676ea6be57209aa9116d7#diff-28707e6fe74fc702b56e27eaff233df0R40 ? (I haven't checked via real client yet)

It's my curiosity. Is this style is common in JS? (Since I'm not familiar with JS.)